File: Rust/bloom_filter/examples/basic.rs
//! Basic usage example for Bloom Filter

use bloom_filter::{BloomConfig, BloomFilter, ScalableBloomFilter};

fn main() {
    println!("=== Bloom Filter Examples ===\n");
    
    // Example 1: Basic usage
    println!("1. Basic Bloom Filter Usage:");
    let mut filter = BloomFilter::<&str>::with_rate(1000, 0.01);
    
    // Insert some words
    let words = vec!["apple", "banana", "cherry", "date", "elderberry"];
    for word in &words {
        filter.insert(word);
        println!("   Inserted: {}", word);
    }
    println!();
    
    // Check membership
    println!("   Membership tests:");
    println!("   'apple' present? {}", filter.contains(&"apple"));
    println!("   'grape' present? {}", filter.contains(&"grape"));
    println!();
    
    // Example 2: Configuration
    println!("2. Custom Configuration:");
    let config = BloomConfig::optimal(10000, 0.001); // 0.1% false positive rate
    println!("   For 10000 items at 0.1% FPR:");
    println!("   - Bit array size: {}", config.size);
    println!("   - Hash functions: {}", config.hash_count);
    println!();
    
    // Example 3: Statistics
    println!("3. Filter Statistics:");
    println!("   Items inserted: {}", filter.len());
    println!("   Bits set: {}", filter.bit_count());
    println!("   Fill ratio: {:.2}%", filter.fill_ratio() * 100.0);
    println!("   Current FPR: {:.4}%", filter.current_false_positive_rate() * 100.0);
    println!();
    
    // Example 4: Scalable Bloom Filter
    println!("4. Scalable Bloom Filter:");
    let mut scalable: ScalableBloomFilter<i32> = ScalableBloomFilter::new(100, 0.01);
    
    // Insert many more items than initial capacity
    for i in 0..5000 {
        scalable.insert(&i);
    }
    
    println!("   Inserted 5000 items (initial capacity: 100)");
    println!("   Internal filters: {}", scalable.filter_count());
    println!("   Contains 100? {}", scalable.contains(&100));
    println!("   Contains 9999? {}", scalable.contains(&9999)); // Not inserted
    println!();
    
    // Example 5: Check and Insert
    println!("5. Check-and-Insert Pattern:");
    let mut filter2 = BloomFilter::<i32>::with_rate(100, 0.01);
    
    let result = filter2.check_and_insert(&42);
    println!("   First insert of 42: {} (was new)", !result);
    
    let result = filter2.check_and_insert(&42);
    println!("   Second insert of 42: {} (already existed)", result);
    println!();
    
    // Example 6: Serialization
    println!("6. Serialization:");
    let mut filter3 = BloomFilter::<&str>::with_rate(100, 0.01);
    filter3.insert(&"persistent");
    filter3.insert(&"storage");
    
    let bytes = filter3.to_bytes();
    println!("   Serialized to {} bytes", bytes.len());
    
    let restored = BloomFilter::<&str>::from_bytes(&bytes).unwrap();
    println!("   Restored filter contains 'persistent': {}", restored.contains(&"persistent"));
    println!("   Restored filter contains 'missing': {}", restored.contains(&"missing"));
    println!();
    
    // Example 7: Merge filters
    println!("7. Merging Filters:");
    let config = BloomConfig::optimal(100, 0.01);
    let mut filter_a: BloomFilter<i32> = BloomFilter::new(config);
    let mut filter_b: BloomFilter<i32> = BloomFilter::new(config);
    
    for i in 0..50 {
        filter_a.insert(&i);
    }
    for i in 50..100 {
        filter_b.insert(&i);
    }
    
    filter_a.merge(&filter_b).unwrap();
    println!("   Merged two filters (50 items each)");
    println!("   Contains 25? {}", filter_a.contains(&25));
    println!("   Contains 75? {}", filter_a.contains(&75));
    println!();
    
    // Example 8: False positive demonstration
    println!("8. False Positive Rate Test:");
    let mut test_filter = BloomFilter::<i32>::with_rate(1000, 0.01);
    
    // Insert 1000 items
    for i in 0..1000 {
        test_filter.insert(&i);
    }
    
    // Test 10000 items not in the set
    let mut false_positives = 0;
    let test_count = 10000;
    for i in 1000..(1000 + test_count) {
        if test_filter.contains(&i) {
            false_positives += 1;
        }
    }
    
    let actual_rate = false_positives as f64 / test_count as f64;
    println!("   Target FPR: 1.00%");
    println!("   Actual FPR: {:.4}%", actual_rate * 100.0);
    println!("   False positives found: {}/{}", false_positives, test_count);
}

File: Zig/bloom_filter/examples/basic.zig
const std = @import("std");
const bloom_filter = @import("bloom_filter");

pub fn main() !void {
    const allocator = std.heap.page_allocator;

    std.debug.print("=== Bloom Filter Demo ===\n\n", .{});

    // Basic Bloom Filter
    std.debug.print("--- Basic Bloom Filter ---\n", .{});
    
    var bf = try bloom_filter.BloomFilter.init(allocator, .{
        .expected_items = 1000,
        .false_positive_rate = 0.01,
    });
    defer bf.deinit();

    std.debug.print("Filter size: {} bits, {} hash functions\n", .{
        bf.bitCount(),
        bf.hashFunctionCount(),
    });

    // Add some fruits
    const fruits = [_][]const u8{ "apple", "banana", "cherry", "date", "elderberry" };
    for (fruits) |fruit| {
        bf.add(fruit);
        std.debug.print("Added: {s}\n", .{fruit});
    }

    std.debug.print("\nItems inserted: {}\n", .{bf.itemCount()});
    std.debug.print("Bits set: {}\n", .{bf.bitsSet()});
    std.debug.print("Estimated FP rate: {:.4}\n", .{bf.estimateFalsePositiveRate()});

    // Check membership
    std.debug.print("\nMembership tests:\n", .{});
    const test_items = [_][]const u8{ "apple", "banana", "fig", "grape", "kiwi" };
    for (test_items) |item| {
        const result = bf.mightContain(item);
        if (result) {
            std.debug.print("  {s}: possibly in set\n", .{item});
        } else {
            std.debug.print("  {s}: definitely NOT in set\n", .{item});
        }
    }

    // Integer operations
    std.debug.print("\n--- Integer Operations ---\n", .{});
    bf.addU64(42);
    bf.addU64(100);
    bf.addU64(999);

    const test_numbers = [_]u64{ 42, 100, 999, 123, 456 };
    for (test_numbers) |num| {
        const result = bf.mightContainU64(num);
        std.debug.print("  {}: {}\n", .{ num, result });
    }

    // Counting Bloom Filter
    std.debug.print("\n--- Counting Bloom Filter (supports removal) ---\n", .{});
    
    var cbf = try bloom_filter.CountingBloomFilter.init(allocator, .{
        .expected_items = 100,
        .false_positive_rate = 0.01,
    });
    defer cbf.deinit();

    cbf.add("red");
    cbf.add("green");
    cbf.add("blue");
    std.debug.print("Added colors: red, green, blue\n", .{});

    std.debug.print("Before removal:\n", .{});
    std.debug.print("  red: {}\n", .{cbf.mightContain("red")});
    std.debug.print("  green: {}\n", .{cbf.mightContain("green")});
    std.debug.print("  blue: {}\n", .{cbf.mightContain("blue")});

    // Remove an item
    if (cbf.remove("green")) {
        std.debug.print("\nRemoved 'green'\n", .{});
    }

    std.debug.print("After removal:\n", .{});
    std.debug.print("  red: {}\n", .{cbf.mightContain("red")});
    std.debug.print("  green: {}\n", .{cbf.mightContain("green")});
    std.debug.print("  blue: {}\n", .{cbf.mightContain("blue")});

    // Scalable Bloom Filter
    std.debug.print("\n--- Scalable Bloom Filter (auto-expands) ---\n", .{});
    
    var sbf = try bloom_filter.ScalableBloomFilter.init(allocator, 0.01);
    defer sbf.deinit();

    std.debug.print("Adding 500 items...\n", .{});
    for (0..500) |i| {
        var buf: [20]u8 = undefined;
        const str = std.fmt.bufPrint(&buf, "user_{}", .{i}) catch unreachable;
        sbf.add(str);
    }

    std.debug.print("Total filters: {}\n", .{sbf.filters.items.len});
    std.debug.print("Total items: {}\n", .{sbf.itemCount()});

    // Test some items
    std.debug.print("\nTesting items:\n", .{});
    std.debug.print("  user_0: {}\n", .{sbf.mightContain("user_0")});
    std.debug.print("  user_100: {}\n", .{sbf.mightContain("user_100")});
    std.debug.print("  user_499: {}\n", .{sbf.mightContain("user_499")});
    std.debug.print("  nonexistent: {}\n", .{sbf.mightContain("nonexistent")});

    // Union operation
    std.debug.print("\n--- Union Operation ---\n", .{});
    
    var bf1 = try bloom_filter.BloomFilter.initWithParams(allocator, 1000, 5);
    defer bf1.deinit();
    
    var bf2 = try bloom_filter.BloomFilter.initWithParams(allocator, 1000, 5);
    defer bf2.deinit();

    bf1.add("cat");
    bf1.add("dog");
    bf2.add("bird");
    bf2.add("fish");

    try bf1.unionWith(&bf2);

    std.debug.print("After union:\n", .{});
    std.debug.print("  cat: {}\n", .{bf1.mightContain("cat")});
    std.debug.print("  dog: {}\n", .{bf1.mightContain("dog")});
    std.debug.print("  bird: {}\n", .{bf1.mightContain("bird")});
    std.debug.print("  fish: {}\n", .{bf1.mightContain("fish")});

    // Serialization
    std.debug.print("\n--- Serialization ---\n", .{});
    
    var bf_orig = try bloom_filter.BloomFilter.init(allocator, .{
        .expected_items = 100,
        .false_positive_rate = 0.01,
    });
    defer bf_orig.deinit();

    bf_orig.add("saved_item_1");
    bf_orig.add("saved_item_2");
    bf_orig.add("saved_item_3");

    const serialized = try bf_orig.serialize(allocator);
    defer allocator.free(serialized);

    std.debug.print("Serialized size: {} bytes\n", .{serialized.len});

    var bf_restored = try bloom_filter.BloomFilter.deserialize(allocator, serialized);
    defer bf_restored.deinit();

    std.debug.print("Restored filter checks:\n", .{});
    std.debug.print("  saved_item_1: {}\n", .{bf_restored.mightContain("saved_item_1")});
    std.debug.print("  saved_item_2: {}\n", .{bf_restored.mightContain("saved_item_2")});
    std.debug.print("  saved_item_3: {}\n", .{bf_restored.mightContain("saved_item_3")});
    std.debug.print("  unknown_item: {}\n", .{bf_restored.mightContain("unknown_item")});

    std.debug.print("\n=== Demo Complete ===\n", .{});
}